pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -245,140 +206,133 @@ func (r *multiCIDRRangeAllocator) Run(stopCh <-chan struct{}) {
 		return
 	}
 
-	// raWaitGroup is used to wait for the RangeAllocator to finish the goroutines.
-	var raWaitGroup sync.WaitGroup
-
 	for i := 0; i < cidrUpdateWorkers; i++ {
-		raWaitGroup.Add(2)
-		go func() {
-			r.worker(stopCh)
-			raWaitGroup.Done()
-		}()
+		go wait.Until(r.nodeWorker, r.workerLoopPeriod, stopCh)
+		go wait.Until(r.cccWorker, r.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
 
-		go func() {
-			r.cccWorker()
-			raWaitGroup.Done()
-		}()
+func (r *multiCIDRRangeAllocator) nodeWorker() {
+	for r.processNextNodeItem() {
+	}
+}
 
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *multiCIDRRangeAllocator) processNextNodeItem() bool {
+	cKey, quit := r.nodeQueue.Get()
+	if quit {
+		return false
 	}
+	defer r.nodeQueue.Done(cKey)
 
-	raWaitGroup.Wait()
+	err := r.syncNode(cKey.(string))
+	r.handleNodeErr(err, cKey)
 
-	<-stopCh
+	return true
 }
 
-func (r *multiCIDRRangeAllocator) worker(stopChan <-chan struct{}) {
-	for {
-		select {
-		case workItem, ok := <-r.nodeCIDRUpdateChannel:
-			r.lock.Lock()
-			if !ok {
-				klog.Error("Channel nodeCIDRUpdateChannel was unexpectedly closed")
-				return
-			}
-			if err := r.updateCIDRsAllocation(workItem); err != nil {
-				// Requeue the failed node for update again.
-				r.nodeCIDRUpdateChannel <- workItem
-			}
-			r.lock.Unlock()
-		case <-stopChan:
-			klog.Infof("MultiCIDRRangeAllocator worker is stopping.")
-			return
-		}
+func (r *multiCIDRRangeAllocator) syncNode(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing node %q. (%v)", key, time.Since(startTime))
+	}()
+
+	node, err := r.nodeLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		// delete the assigned CIDR to that node
+		// TODO, we may have to use a cache if we don't have this info already
```

Comment:
this is the main downside with this pattern, so the object doesn't exist but you have to have the information that you need to clean on its absence ... in this case maybe we have something in the cidrset, if not, we have to be able to delete all resources there depending on the nodename